File: utils_cpp/list.cpp
#include "list.h"

bool is_equal_list(ListNode *l1, ListNode *l2)
{
    while (true) {
        if (!l1 && !l2)
            return true;
        if (!l1 || !l2)
            return false;
        if (l1->val != l2->val)
            return false;
        l1 = l1 ? l1->next : l1;
        l2 = l2 ? l2->next : l2;
    }
}

File: util_go/tree/list.go
package list

type ListNode struct {
	Val int
	Next *ListNode
}

func BuildList(s []int) *ListNode {
	preHead := &ListNode{}
	p := preHead
	for n := range s {
		p.Next = *ListNode{Val: n}
		p = p.Next
	}
	return preHead.Next
}

func IsEqualList(l1, l2 *ListNode) bool {
	if l1 == nil && l2 == nil {
		return true
	}
	if l1 == nil || l2 == nil {
		return false
	}
	if l1.Val != l2.Val {
		return false
	}
	return IsEqualList(l1.Next, l2.Next)
}
